Refactor dispersion plot

dlukes opened the pull request:
This is a follow-up to a suggestion I made a while ago in #2247. I use dispersion plots a lot when teaching beginners. Students are often impressed and ask if tweaks can be made to the result. So I introduce them to the `??` IPython magic, which shows them the source code, but we typically stop short of actually playing around with a few tweaks, because:

- The data preparation part is written in a functional style which is a bit inscrutable, it emphasizes Python features which aren't typically the bread and butter of beginner introductions to Python.
- The plotting part uses Matplotlib's state-machine API, instead of the object-oriented one which is currently favored.

With the changes I'm proposing here, my hope is to make dispersion plot more hackable for beginners. I tried breaking up the changes into individual commits to facilitate discussion around different aspects of the issue. I'm aware that people's experience with teaching beginners may vary, so I'm very much open to discussion.

One part I'd definitely like to keep though is that `dispersion_plot` should return the `Axes` object it creates, so that the first tweaking attempts can be made without even looking at the source code. In light of #2239, I hope this is relatively uncontroversial.

dlukes committed:
Modernize Matplotlib code for dispersion plot

- Use object-oriented instead of state-machine interface.
- Return Axes object to allow additional customization (see #2239).
- Remove useless scalex kwarg (it's supposed to be a bool and True by
  default, so passing in 0.1 is confusing).

dlukes committed:
Use default palette in dispersion plot

dlukes committed:
Refactor data preparation in dispersion plot

Make the code a bit more concise and readable for beginners, who may
want to use it as a starting point for their own tweaked dispersion
plot.

(Incidentally, this version is also a bit faster since it replaces the
nested loop over words with the in operator on a dict, but that's not
the main goal.)

dlukes committed:
Casefold instead of lower in dispersion plot

str.casefold is the method primarily meant for caseless comparison.

dlukes committed:
Dispersion plot docstring tweak

dlukes commented on nltk/draw/dispersion.py:
```diff
@@ -15,51 +15,47 @@ def dispersion_plot(text, words, ignore_case=False, title="Lexical Dispersion Pl
 
     :param text: The source text
-    :type text: list(str) or enum(str)
+    :type text: list(str) or iter(str)
```
`enum` in this context doesn't really make sense to me, but maybe that's my fault?

tomaarsen replied:
No, you're right, `enum` was always an odd choice here, as `text` is immediately converted to a list.

dlukes commented on nltk/draw/dispersion.py:
```diff
@@ -15,51 +15,47 @@ def dispersion_plot(text, words, ignore_case=False, title="Lexical Dispersion Pl
         ) from e
 
-    text = list(text)
-    words.reverse()
```
Reversing in place can lead to surprises when the same list of words is used in multiple dispersion plots (or other places besides a dispersion plot).

tomaarsen replied:
Agreed.

tomaarsen committed:
Reraise ImportError if importing matplotlib fails

Rather than ValueError.
Additionally, add a space between "... installed." and "See ..."

tomaarsen committed:
Add docstring for return value to dispersion plot

tomaarsen commented:
This is looking great: a collection of each very reasonable changes. I'll happily merge this when the tests go green. 
I also modified the `ValueError` that is thrown if someone does not have `matplotlib` into an `ImportError`. Furthermore, I added a simplel docstring to show that the function returns an Axes object (now). In doing so I accidentally pulled in some commits from the develop branch, so I had to rebase this PR. My apologies.

Happy teaching!

tomaarsen commented:
The current test failures are due to an outdated cache of the NLTK data. Just today we added BCP47 via #3060 and the corresponding doctests will fail until the NLTK data cache refreshes. Sadly, I can only control the cache on the develop branch (which has the updated NLTK data cache), and not on any PR.
That said, the tests pass for me locally, so I'll just merge this and make sure they still pass on `develop`.

dlukes commented:
Thank you for the quick review and merge!
